### PR TITLE
stream/hexdump: add hexdump stream to dump binary to syslog 

### PR DESCRIPTION
**Pull request by anchao**
Screenshot from 2023-05-24 14-38-21

## Summary

stream/hexdump: add hexdump stream to dump binary to syslog 

Signed-off-by: chao an <anchao@xiaomi.com>

## Impact

N/A

## Testing

ci-check

### CODE REVIEW

File: libs/libc/stream/lib_hexdumpstream.c [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+{
+  FAR struct lib_hexdumpstream_s *rthis = (FAR void *)this;
+
+  if (rthis->pending > 0)
+    {
+      rthis->buffer[rthis->pending] = '\0';
+
+      if (rthis->backend == NULL)
+        {
+          syslog(rthis->priority, "%s\n", rthis->buffer);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_hexdumpstream.c:92]: let's decouple syslog from hexdumpstream

Comment (anchao) [reply libs/libc/stream/lib_hexdumpstream.c:92]: Done

File: libs/libc/stream/lib_hexdumpstream.c [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <syslog.h>
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_hexdumpstream.c:27]: remove, don't need anymore

File: libs/libc/stream/lib_hexdumpstream.c [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+ * Name: hexdumpstream_flush
+ ****************************************************************************/
+
+static int hexdumpstream_flush(FAR struct lib_outstream_s *this)
+{
+  FAR struct lib_hexdumpstream_s *rthis = (FAR void *)this;
+
+  if (rthis->pending > 0)
+    {
+      rthis->buffer[rthis->pending] = '\0';
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_hexdumpstream.c:88]: don't need zero

File: include/nuttx/streams.h [hunk 4]
```diff
@@ -225,6 +225,14 @@ struct lib_bufferedoutstream_s
   char                        buffer[CONFIG_STREAM_OUT_BUFFER_SIZE];
 };
 
+struct lib_hexdumpstream_s
+{
+  struct lib_outstream_s      public;
+  FAR struct lib_outstream_s *backend;
+  int                         pending;
+  char                        buffer[CONFIG_STREAM_HEXDUMP_BUFFER_SIZE + 1];
```
Comment (xiaoxiang781216) [inline include/nuttx/streams.h:233]: remove + 1

File: libs/libc/stream/lib_hexdumpstream.c [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <syslog.h>
+
+#include "libc.h"
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_hexdumpstream.c:29]: why need

File: libs/libc/stream/lib_hexdumpstream.c (hunk 2 above)
Comment (anchao) [reply libs/libc/stream/lib_hexdumpstream.c]: Done

File: include/nuttx/streams.h (hunk 4 above)
Comment (anchao) [reply include/nuttx/streams.h]: Done

File: libs/libc/stream/lib_hexdumpstream.c (hunk 3 above)
Comment (anchao) [reply libs/libc/stream/lib_hexdumpstream.c]: Done

File: libs/libc/stream/lib_hexdumpstream.c (hunk 5 above)
Comment (anchao) [reply libs/libc/stream/lib_hexdumpstream.c]: Done

**Comment by anchao** [comment]
> please add a commit message

I added a diagram to describe the feature of Hexdump stream

File: libs/libc/stream/lib_hexdumpstream.c [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+/****************************************************************************
+ * Name: hexdumpstream_putc
+ ****************************************************************************/
+
+static void hexdumpstream_putc(FAR struct lib_outstream_s *this, int ch)
+{
+  FAR struct lib_hexdumpstream_s *rthis = (FAR void *)this;
+  int outlen = sizeof(rthis->buffer);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_hexdumpstream.c:101]: change sizeof(rthis->buffer) to CONFIG_STREAM_HEXDUMP_BUFFER_SIZE

File: libs/libc/stream/lib_hexdumpstream.c [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+/****************************************************************************
+ * Name: hexdumpstream_puts
+ ****************************************************************************/
+
+static int hexdumpstream_puts(FAR struct lib_outstream_s *this,
+                           FAR const void *buf, int len)
+{
+  FAR struct lib_hexdumpstream_s *rthis = (FAR void *)this;
+  const unsigned char *p = buf;
+  int outlen = sizeof(rthis->buffer);
```
Comment (xiaoxiang781216) [inline libs/libc/stream/lib_hexdumpstream.c:124]: change sizeof(rthis->buffer) to CONFIG_STREAM_HEXDUMP_BUFFER_SIZE

File: libs/libc/stream/lib_hexdumpstream.c (hunk 6 above)
Comment (anchao) [reply libs/libc/stream/lib_hexdumpstream.c]: Done

File: libs/libc/stream/lib_hexdumpstream.c (hunk 7 above)
Comment (anchao) [reply libs/libc/stream/lib_hexdumpstream.c]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by davids5** [comment]
@anchao - this is really a nice piece of work, Thank you for it.

**Pull request merged**